### PR TITLE
Extension systems + Relative Extension systems

### PR DESCRIPTION
**Pull request by TOTBWF**
# Description

This PR defines extension systems and relative extension systems, and proves their equivalence with monads.

## Checklist

Before submitting a merge request, please check the items below:

- [x] I've read [the contributing guidelines](https://github.com/plt-amy/1lab/blob/main/CONTRIBUTING.md).
- [x] The imports of new modules have been sorted with `support/sort-imports.hs` (or `nix run --experimental-features nix-command -f . sort-imports`).
- [x] All new code blocks have "agda" as their language.

If your change affects many files without adding substantial content, and
you don't want your name to appear on those pages (for example, treewide
refactorings or reformattings), start the commit message and PR title with `chore:`.

### CODE REVIEW

**Comment by Lavenza** [comment]
[Pull request preview](https://preview.1lab.dev/342)

File: src/Cat/Diagram/Monad/Relative.lagda.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+By definition, [[monads]] must be endofunctors. For most applications, this
+is not relevant, and the theory of monads is quite rich. However, this
+restriction can become problematic. For simplicity, let $T$ be a monad
+on set interpreted as an algebraic theory, where $T(A)$ is the syntax of
+$T$ with $A$-many variables, the unit constructs a term from a variable,
+and the join performs substitution. If we wish to give a more refined
+treatment of variables^[for instance, if a term can only have finitely
+many variables], then we immediately run into an issue: $T$ can no
+longer an endofunctor $\Sets \to \Sets$, and restricting $T$ to
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Relative.lagda.md:30]:
```suggestion
longer be an endofunctor $\Sets \to \Sets$, and restricting $T$ to
```

File: src/Cat/Diagram/Monad/Relative.lagda.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+$\FinSets \to \FinSets$ is also out of the question, as the collection
+of syntax trees will not be finite.
+
+If we wish to capture such situations, then we will need to generalize
+the notion of monad. The fundamental problem with doing so is that we
+cannot iterate a functor $M : \cJ \to \cC$, so themost natural starting
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Relative.lagda.md:36]:
```suggestion
cannot iterate a functor $M : \cJ \to \cC$, so the most natural starting
```

File: src/Cat/Diagram/Monad/Relative.lagda.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+The functorial action of $M$ on $f$ can be recovered by extending
+$\eta \circ f$.
+
+```agda
+    M₁ : ∀ {x y} → J.Hom x y → C.Hom (M₀ x) (M₀ y)
+    M₁ f = bind (unit C.∘ F.₁ f)
+```
+
+Furthermore, the latter two equations ensure naturality of the unit and
+extension, resp.
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Relative.lagda.md:95]:
```suggestion
extension, respectively.
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+2. A family of morphisms $\eta_X : \cC(X, M(X))$, called the **unit**
+of the extension system
+3. An extension operation $(-)^{M} : \cC(X,M(y)) \to \cC(M(X), M(Y))$.
+  Gesturing towards the "monads" found in functional programming, we shall
+  call this operation `bind`.
+
+Note that we do not require the mapping of objects to be functorial, nor
+the do we require the unit to be natural.
+
+We impose 3 additional equations atop this structure[^contrast this with
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:55]:
```suggestion
We impose 3 additional equations atop this structure^[contrast this with
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    monad .mult-assoc =
+      E.bind id ∘ E.bind (E.unit ∘ E.bind id) ≡⟨ E.bind-∘ _ _ ⟩
+      E.bind (E.bind id ∘ E.unit ∘ E.bind id) ≡⟨ ap E.bind (cancell (E.bind-unit-∘ id) ∙ sym (idr _)) ⟩
+      E.bind (E.bind id ∘ id)                 ≡˘⟨ E.bind-∘ _ _ ⟩
+      E.bind id ∘ E.bind id                   ∎
+```
+
+Moreover, these two functions constitute an equivalence between monads and
+extension systems. In light of this fact, we will silently identify monads
+and extension algebras whenever it is convenient.
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:190]:
```suggestion
and extension systems whenever it is convenient.
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+      (α.ν ∘ join) ∘ bind (unit ∘ M₁ f ∘ g)    ≡⟨ pushl (sym $ α.ν-mult) ⟩
+      α.ν ∘ M₁ α.ν ∘ bind (unit ∘ M₁ f ∘ g)    ≡⟨ ap (α.ν ∘_) (bind-naturall _ _) ⟩
+      α.ν ∘ bind ⌜ M₁ α.ν ∘ unit ∘ M₁ f ∘ g ⌝  ≡⟨ ap! (centralizel (sym $ unit-natural _)) ⟩
+      α.ν ∘ bind (unit ∘ (α.ν ∘ M₁ f) ∘ g)     ∎
+```
+</details>
+
+Conversely, a monad algebra over $E$ can be derived from an extension
+algebra $\nu : \cC(A, X) \to \cC(M(A), X)$ over $E$ by restricting to
+$\nu(id_{X}) : \cC(M(X), X)$.
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:321]:
```suggestion
$\nu(\id{X}) : \cC(M(X), X)$.
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+Extension-system C = Relative-extension-system (Id {C = C})
+
+module Extension-system {o ℓ} {C : Precategory o ℓ} (E : Extension-system C) where
+  open Cat.Reasoning C
+  open Relative-extension-system E public
+```
+
+
+We can recover the join of the monad by extending the identity morphism
+$id_{M(X)} : \cC(M(X), M(X))$.
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:79]:
```suggestion
$\id{M(X)} : \cC(M(X), M(X))$.
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+  call this operation `bind`.
+
+Note that we do not require the mapping of objects to be functorial, nor
+the do we require the unit to be natural.
+
+We impose 3 additional equations atop this structure[^contrast this with
+the 7 equations required for a monad: 2 for functoriality, 2 for naturality,
+and 3 for unitality/associativity]:
+
+1. For every $X : \cC$, $(\eta_X)^M = id_{M(X)}$
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:59]:
```suggestion
1. For every $X : \cC$, $(\eta_X)^M = \id{M(X)}$
```

**Review by plt-amy — CHANGES_REQUESTED** [review]
Sadly GitHub won't let me submit my review as [a patch](https://gist.github.com/plt-amy/6115e2a92facb4a7ee54239e962e7bb4).

File: src/Cat/Diagram/Monad.lagda.md [hunk 9]
```diff
@@ -372,3 +373,42 @@ $\cC^M$.
     Free⊣Forget .zig = ext left-ident
     Free⊣Forget .zag {x} = x .snd .ν-unit
 ```
+
+<!--
+```
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad.lagda.md:378]:
```suggestion
```agda
```

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+open _=>_
+```
+-->
+
+```agda
+module Cat.Diagram.Monad.Extension where
+```
+
+<!--
+```
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:19]:
```suggestion
```agda
```

**Review by ncfavier — APPROVED** [review]

**Comment by ncfavier** [comment]
I'm going to guess the auto-merge wasn't intentional given the timing and disable it, but this might as well be merged as far as I'm concerned.

File: src/Cat/Diagram/Monad/Extension.lagda.md [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+module Cat.Diagram.Monad.Extension where
+```
+
+<!--
+```agda
+module _ {o ℓ} (C : Precategory o ℓ) where
+```
+-->
+
+# Extension Systems {defines=extension-system}
```
Comment (ncfavier) [inline src/Cat/Diagram/Monad/Extension.lagda.md:24]:
```suggestion
# Extension systems {defines=extension-system}
```

**Review by ncfavier — APPROVED** [review]

**Comment by TOTBWF** [comment]
I think we need @plt-amy to sign off to appease CI

**Review by plt-amy — APPROVED** [review]

**Pull request merged**